### PR TITLE
msgr2: unit test infrastructure and reconnect race fix

### PR DESCRIPTION
**Pull request by rjfd**
This PR adds "interceptor" calls to the msgr2 protocol code so that we have a fine grain control over the protocol steps when write unit tests.

This PR also adds a fix for reconnect races, which happens when two peers are reconnecting to a previous existing session.

Signed-off-by: Ricardo Dias <rdias@suse.com>

### CODE REVIEW

**Comment by rjfd** [comment]
First monthrash suite pass looks good:
http://pulpito.ceph.com/rdias-2019-02-20_14:50:54-rados:monthrash-wip-rdias2-testing-distro-basic-smithi/

**Comment by rjfd** [comment]
rados suite also looks good:

http://pulpito.ceph.com/rdias-2019-02-21_07:23:11-rados-wip-rdias2-testing-distro-basic-smithi/

File: CMakeLists.txt [hunk 1]
```diff
@@ -576,6 +576,7 @@ option(ENABLE_COVERAGE "Coverage is enabled" OFF)
 option(PG_DEBUG_REFS "PG Ref debugging is enabled" OFF)
 
 option(WITH_TESTS "enable the build of ceph-test package scripts/binaries" ON)
+set("UNIT_TESTS_BUILT", ${WITH_TESTS})
```
Comment (liewegas) [inline CMakeLists.txt]: are release packages built with ``WITH_TESTS=off``?

Comment (rjfd) [reply CMakeLists.txt]: that's a good question, who can we ping to answer that?

**Pull request merged**